parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java

```diff
@@ -0,0 +1,114 @@
+    private Calendar startTime;
+    private final UpdateEngagementIntervalCalculator intervalCalculator;
+
+    public EngagementManager(ParselyTracker parselyTracker, Timer parentTimer, long intervalMillis, Map<String, Object> baseEvent, UpdateEngagementIntervalCalculator intervalCalculator) {
+        this.parselyTracker = parselyTracker;
+        this.baseEvent = baseEvent;
+        this.parentTimer = parentTimer;
+        this.intervalCalculator = intervalCalculator;
+        latestDelayMillis = intervalMillis;
+        totalTime = 0;
+        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+    }
+
+    public boolean isRunning() {
+        return started;
+    }
+
+    public void start() {
+        scheduleNextExecution(latestDelayMillis);
+        started = true;
+        startTime = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+    }
+
+    public void stop() {
+        waitingTimerTask.cancel();
+        started = false;
+    }
+
+    public boolean isSameVideo(String url, String urlRef, ParselyVideoMetadata metadata) {
+        Map<String, Object> baseMetadata = (Map<String, Object>) baseEvent.get("metadata");
+        return (baseEvent.get("url").equals(url) &&
+                baseEvent.get("urlref").equals(urlRef) &&
+                baseMetadata.get("link").equals(metadata.link) &&
+                (int) (baseMetadata.get("duration")) == metadata.durationSeconds);
+    }
+
+    private void scheduleNextExecution(long delay) {
+        TimerTask task = new TimerTask() {
+            public void run() {
+                doEnqueue(scheduledExecutionTime());
+                latestDelayMillis = intervalCalculator.updateLatestInterval(startTime);
+                scheduleNextExecution(latestDelayMillis);
+            }
+
+            public boolean cancel() {
+                boolean output = super.cancel();
+                // Only enqueue when we actually canceled something. If output is false then
+                // this has already been canceled.
+                if (output) {
+                    doEnqueue(scheduledExecutionTime());
+                }
+                return output;
+            }
+        };
+        latestDelayMillis = delay;
+        parentTimer.schedule(task, delay);
+        waitingTimerTask = task;
+    }
+
+    private void doEnqueue(long scheduledExecutionTime) {
+        // Create a copy of the base event to enqueue
+        Map<String, Object> event = new HashMap<>(baseEvent);
+        ParselyTracker.PLog(String.format("Enqueuing %s event.", event.get("action")));
+
+        // Update `ts` for the event since it's happening right now.
+        Calendar now = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        @SuppressWarnings("unchecked")
+        Map<String, Object> baseEventData = (Map<String, Object>) event.get("data");
+        assert baseEventData != null;
+        Map<String, Object> data = new HashMap<>((Map<String, Object>) baseEventData);
```

Comment:
Minor (🔍): Consider removing the redundant `(Map<String, Object>)` casting for `baseEventData`.

Warning: `Casting 'baseEventData' to 'Map<String, Object>' is redundant`